### ISSUE TITLE
HTTP digest as public API

### ISSUE THREAD

**Issue by MrSpider** [civetweb/civetweb#426]
Hello,

Would it be possible to make the methods for digest authentication a public API?
We authenticate clients in civetweb first by IP and then for some IPs by digest auth.

The methods needed for digest are:
[int check_authorization(struct mg_connection *conn, const char *path)](https://github.com/civetweb/civetweb/blob/master/src/civetweb.c#L6753)
[void send_authorization_request(struct mg_connection *conn)](https://github.com/civetweb/civetweb/blob/master/src/civetweb.c#L6802)

If it's ok with you, I would make a pull request for a public API for these methods.
Which would something like this:

```
int mg_handle_digest_authorization(struct mg_connection *conn, const char *path)
{
        if (!check_authorization(conn, path)) {
            send_authorization_request(conn);
            return 1;
        }
        return 0;
}
```

**Comment by bel2125** [comment]
In `civetweb.h`, we have the `mg_authorization_handler` callback function, set for an uri-pattern using `mg_set_auth_handler`.

In the auth_handler callback, you could first check the IP address, some cookies or whatever. 
For digest authentication, you would need `mg_send_digest_authorization_request` (basically `send_authorization_request` with an `mg_` prefix and some return value, and maybe some parameter for "digest"). And something similar for `mg_check_digest_authorization`. Probably called  `_authentication` instead of `_authorization`, like in the RFC (https://tools.ietf.org/html/rfc2617).

Before doing any pull request, we should discuss all API enhancements here.

**Comment by MrSpider** [comment]
I think it may be better to just change the name instead if adding a extra parameter for the auth type.
`mg_send_digest_authorization_request(...)`  instead of `mg_send_authorization_request(struct mg_connection *, <sometype>)`.

I was looking into whether the path should be passed in `mg_check_digest_authentication`. If not `interpret_uri` would have to be called again.

When e.g. basic auth is added a new API function could be added which takes the type of auth to use?

```
/* Checks if the request is authenticated using HTTP digest authentication.
   Returns 1 if the request is authenticated, 0 otherwise. */
CIVETWEB_API int
mg_check_digest_authentication(struct mg_connection *conn, const char *path);


/* Sends the HTTP digest authorization request. */
CIVETWEB_API void
mg_send_digest_authorization_request(struct mg_connection *conn);
```

**Comment by bel2125** [comment]
> I think it may be better to just change the name instead if adding a extra parameter for the auth type.
mg_send_digest_authorization_request(...) instead of mg_send_authorization_request(struct mg_connection *, <sometype>)

I meant parameters for the digest authorization itself, such as the auth domain, maybe the nonce.

I would not add the path explicitly, this should be already in `mg_get_request_info(conn)->local_uri` (where mg_get_request_info is a simple inexpensive getter). This is the way, all other functions work as well.
And you already specified a path (uri) when calling `mg_set_auth_handler(struct mg_context *ctx, const char *uri, mg_request_handler handler, void *cbdata)`)

**Comment by bel2125** [comment]
The following functions have been added to [the public API `civetweb.h`](https://github.com/civetweb/civetweb/blob/46dc39cc4fbc8f9e5b1a673dc416bdd7ad5f8106/include/civetweb.h#L776):

```C
CIVETWEB_API int
mg_send_digest_access_authentication_request(struct mg_connection *conn,
                                             const char *realm);


CIVETWEB_API int
mg_check_digest_access_authentication(struct mg_connection *conn,
                                      const char *realm,
                                      const char *filename);
```
The additional `realm` parameter can be NULL to use the `authentication_domain` parameter from mg_start, but it also allows to use multiple realms on the same server instance (in callbacks).

**Comment by bel2125** [comment]
The new API is released in V1.10 now.
